### PR TITLE
Moving release process to Bintray 

### PR DESCRIPTION
**Pull request by langsmith**
This pr resolves https://github.com/mapbox/mapbox-java/issues/995 by moving the Java SDK artifact release process to Bintray. The work in this pr is largely based off of [similar work done in `mapbox-gl-native`](https://github.com/mapbox/mapbox-gl-native/pull/13999) and discussion that's happened in https://github.com/mapbox/mapbox-java/issues/995.

### CODE REVIEW

**Comment by langsmith** [comment]
I believe the only remaining work on this pr is figuring out how to delineate whether a version is a SNAPSHOT or a stable/alpha version as mentioned by @LukasPaczos at https://github.com/mapbox/mapbox-java/issues/995#issuecomment-490910570

**Comment by langsmith** [comment]
> figuring out how to delineate whether a version is a SNAPSHOT or a stable/alpha version

https://github.com/mapbox/mapbox-java/pull/1037/commits/9f5db8378e4bf032f9d56e9ff329b7f520f96a66 added logic to `circle.yml` to determine this.

**Comment by langsmith** [comment]
@LukasPaczos , I need to request OSS Jfrog access for SNAPSHOTs of each of the 4 Java packages https://bintray.com/message/addPackageToJCenter?pkgPath=%2Fmapbox%2Fmapbox%2Fmapbox-sdk-turf&tab=general 👇 
Screen Shot 2019-05-17 at 4 59 05 PM


When you did this for the Maps SDK, do you remember what you checked/entered?

I tried 👇 
Screen Shot 2019-05-17 at 5 01 09 PM

However, that gives me the following error in the red box. Have you seen this before? What was your work around, if any?

Screen Shot 2019-05-17 at 5 01 56 PM

**Comment by langsmith** [comment]
Or can I make up some sort of unique ID such as `com.mapbox.mapboxsdk.mapbox-sdk-turf`?

https://www.jfrog.com/confluence/display/RTF/Deploying+Snapshots+to+oss.jfrog.org#DeployingSnapshotstooss.jfrog.org-GettingStartedwithOJO

**Comment by LukasPaczos** [comment]
I did those exact steps while setting up Maps SDK. It's weird that you wouldn't be able to publish another artifact under the same Group ID to Artifactory. Could you try pushing them anyway? [The package structure is established on JCenter](http://oss.jfrog.org/oss-snapshot-local/com/mapbox/mapboxsdk/), so maybe it'll pick it up?

If not, the best way would be to contact support - maybe that's not an expected use-case.

**Comment by langsmith** [comment]
Hm, ok. Will look into it @LukasPaczos . Contacted support as backup.

**Comment by LukasPaczos** [comment]
Any progress here @langsmith? Have you tried pushing the `SNAPSHOT`? Any news from the support?

**Comment by langsmith** [comment]
> news from the support

No news, probably because we're not Enterprise customers, so there's no urgency on their part.

> Could you try pushing them anyway
> Have you tried pushing the SNAPSHOT

You mean publishing the test SNAPSHOT as if it were a stable release, correct?

**Comment by langsmith** [comment]
>Could you try pushing them anyway
> Have you tried pushing the SNAPSHOT

As a cautious first step I just did a test run of the 4 modules. Then I published `4.9.0-SNAPSHOT` of only Turf rather than all 4 modules: https://bintray.com/mapbox/mapbox/mapbox-sdk-turf/4.9.0-SNAPSHOT

When I clicked on `Click here to get it included.` link 
Screen Shot 2019-05-24 at 3 05 09 PM

I got the following error message

Screen Shot 2019-05-24 at 3 05 02 PM

**Comment by langsmith** [comment]
This is the Turf POM file

Screen Shot 2019-05-24 at 3 07 35 PM

**Comment by langsmith** [comment]
Think I found at least one error with the POM file. Looks like the Turf POM file above incorrectly has `services-geojson` for the geojson dependency artifact id. Based on https://search.maven.org/search?q=Mapbox, it should be `mapbox-sdk-geojson`.

**Comment by langsmith** [comment]
Ok, I adjusted so that the correct `artifact-id` is added to the POM file.
https://github.com/mapbox/mapbox-java/blob/1cc1392edc1f62ea298a549c23b706f81ef4750b/gradle/gradle-bintray.gradle#L13


Deleted all other Java SDK files from Bintray, built release files on my local machine and then re-published to Bintray. Better Turf POM file is created:

Screen Shot 2019-05-24 at 3 45 16 PM

`services` module POM file:

Screen Shot 2019-05-24 at 3 49 13 PM

But darn. Despite all of this, still getting the error message.

Screen Shot 2019-05-24 at 3 50 30 PM

**Comment by langsmith** [comment]
Do the POM files need a `<packaging>jar</packaging>` line? Doesn't seem to be in https://search.maven.org/artifact/com.mapbox.mapboxsdk/mapbox-sdk-services/4.8.0/jar

**Comment by langsmith** [comment]
So weird. 

https://search.maven.org/artifact/com.mapbox.mapboxsdk/mapbox-sdk-services/4.8.0/jar compared with the `services` POM file above (found at https://bintray.com/mapbox/mapbox/download_file?file_path=com%2Fmapbox%2Fmapboxsdk%2Fmapbox-sdk-services%2F4.9.0-SNAPSHOT%2Fmapbox-sdk-services-4.9.0-SNAPSHOT.pom) , reveals no major differences. The name, description, and url tags are in a different location. The bintray POM doesn't have the test scope dependencies either. Not sure whether any of those would lead to the red Bintray error message.

**Comment by LukasPaczos** [comment]
> You mean publishing the test SNAPSHOT as if it were a stable release, correct?

No, I meant publishing a `SNAPSHOT` to an actual `SNAPSHOT` repository to test possible JCenter compatibility ([publishing to artifactory job](https://github.com/mapbox/mapbox-java/pull/1037/files#diff-49e0bb33be7766b39a755e0ae6f81b39R73-R85)). The POM does look fine, and I see the "4.9.0-SNAPSHOT" marked as released https://bintray.com/mapbox/mapbox/mapbox-sdk-turf/4.9.0-SNAPSHOT, it also seems to be resolvable by Gradle (well, it's missing `mapbox-sdk-geojson`, but the `mapbox-sdk-turf` looks fine). Did you succeed after all?

**Comment by langsmith** [comment]
No news from Bintray support. I don't expect any to arrive.

Tried `make build-release` and then `make publish-snapshot`. Got the following 401 for each of the 4 modules:

Screen Shot 2019-05-29 at 8 27 25 PM

I haven't tried accessing the SNAPSHOT via an actual project though.

**Comment by LukasPaczos** [comment]
Since we are still blocked on the JCenter front, I'd suggest removing all `SNAPSHOT` changes and focus on the MavenCentral release flow. We can still publish snapshots to Sonatype and revisit the setup when we know more about the JCenter repositories. What do you think @langsmith?

**Comment by langsmith** [comment]
I was testing with `4.9.0-SNAPSHOT`. The `-SNAPSHOT` ending isn't allowed in the pom file 🤦‍♂ 

https://central.sonatype.org/pages/requirements.html:
Screen Shot 2019-07-01 at 1 20 59 PM

When I switched to `4.9.0` for a **_test_** run of a new release, the MavenCentral request successfully went through:
Screen Shot 2019-07-01 at 1 19 44 PM
Screen Shot 2019-07-01 at 1 16 28 PM

Now we wait for the Bintray team to accept my request so that we have the option to sync the 4 modules' files with MavenCentral 🤞 I won't actually sync because this wasn't a true `4.9.0` release.

**Comment by langsmith** [comment]
> Now we wait for the Bintray team to accept my request so that we have the option to sync the 4 modules' files with MavenCentral

They gave us permission and we're now able to sync files with MavenCentral. SNAPSHOTs also look setup in our Bintray account.  This pr is ready for review.

File: Makefile [hunk 1]
```diff
@@ -27,18 +27,25 @@ javadoc:
...
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-turf:bintrayUpload ; \
+
+publish-snapshot:
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-core:artifactoryPublish ; \
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-geojson:artifactoryPublish ; \
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services:artifactoryPublish ; \
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-turf:artifactoryPublish ; \
+
 
 publish-local:
```
Comment (LukasPaczos) [inline Makefile]: This won't work and can be removed.

File: build.gradle [hunk 2]
```diff
@@ -3,6 +3,7 @@ buildscript {
 
   repositories {
     maven { url 'https://plugins.gradle.org/m2' }
+    maven { url 'https://oss.jfrog.org/artifactory/oss-snapshot-local/' }
```
Comment (LukasPaczos) [inline build.gradle]: This is not needed and can be removed.

File: circle.yml [hunk 3]
```diff
@@ -58,20 +58,23 @@ jobs:
...
       - run:
           name: Build Java libraries
           command: make build-release
       - deploy:
-          name: Publish Java Services To Maven Central
+          name: Publish Java libraries to Bintray
           command: |
-            if [ "${CIRCLE_BRANCH}" == "master" ]; then
-            make publish ;
+            if [[ $CIRCLE_BRANCH == master ]] || [[ $CIRCLE_BRANCH == release-* ]] || [[ $CIRCLE_TAG == android-v* ]]; then
```
Comment (LukasPaczos) [inline circle.yml]: Tag convention for this repo is `v[number]`, others can be removed (besides `master`).

File: services-core/gradle.properties [hunk 4]
```diff
@@ -1,5 +1,5 @@
 VERSION_NAME=4.9.0-SNAPSHOT
 POM_ARTIFACT_ID=mapbox-sdk-core
 POM_NAME=Mapbox Services SDK
```
Comment (LukasPaczos) [inline services-core/gradle.properties]: This seems unused.

File: circle.yml (hunk 3 above)
Comment (langsmith) [reply circle.yml]: Fixed

File: services-core/gradle.properties (hunk 4 above)
Comment (langsmith) [reply services-core/gradle.properties]: Removed `POM_NAME` from each `gradle.properties` file

File: build.gradle (hunk 2 above)
Comment (langsmith) [reply build.gradle]: Removed

File: Makefile (hunk 1 above)
Comment (langsmith) [reply Makefile]: Removed

**Comment by langsmith** [comment]
> Also, the readme should be updated with the new snapshot repository URL.

Updated

**Comment by langsmith** [comment]
>The release process is missing the step that changes the version name and removes the SNAPSHOT reference in order to support a full tag release. Check out this step. This should be applied for each artifact.

Added the step to `circle.yml` and wrote it the way that I did because I deleted `VERSION_NAME` from each nested `gradle.properties` file. The `VERSION_NAME` is the same for all 4 modules/artifacts and it's already declared in the top level `gradle.properties` file.

**Comment by langsmith** [comment]
Ok @LukasPaczos , pushed a commit that addressed your latest review.

File: circle.yml [hunk 5]
```diff
@@ -58,20 +58,29 @@ jobs:
...
       - run:
           name: Build Java libraries
           command: make build-release
       - deploy:
-          name: Publish Java Services To Maven Central
+          name: Publish Java libraries to Bintray
           command: |
-            if [ "${CIRCLE_BRANCH}" == "master" ]; then
-            make publish ;
+            if [[ $CIRCLE_BRANCH == master ]] || [[ $CIRCLE_TAG == android-v* ]]; then
```
Comment (LukasPaczos) [inline circle.yml]: `[[ $CIRCLE_TAG == android-v* ]]` should be `[[ $CIRCLE_TAG == v* ]]`

File: build.gradle [hunk 6]
```diff
@@ -97,12 +102,14 @@ subprojects { subproject ->
   }
 
   if (RELEASE_MODULES.contains(subproject.name)) {
-    subproject.apply from: "${rootDir}/gradle/mvn-push.gradle"
+    subproject.apply plugin: 'com.jfrog.bintray'
```
Comment (LukasPaczos) [inline build.gradle]: Already applied in the `gradle-bintray.gradle`.

File: build.gradle [hunk 7]
```diff
@@ -97,12 +102,14 @@ subprojects { subproject ->
...
+    subproject.apply plugin: "com.jfrog.artifactory"
```
Comment (LukasPaczos) [inline build.gradle]: Same as above.

File: build.gradle [hunk 8]
```diff
@@ -97,12 +102,14 @@ subprojects { subproject ->
...
+    subproject.apply from: "${rootDir}/gradle/artifact-settings.gradle"
```
Comment (LukasPaczos) [inline build.gradle]: Same as above.

File: Makefile [hunk 9]
```diff
@@ -27,18 +27,17 @@ javadoc:
...
-	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-turf:uploadArchives ; \
-
-publish-local:
-	# This publishes to ~/.m2/repository/com/mapbox/mapboxsdk
-	export IS_LOCAL_DEVELOPMENT=true; ./gradlew :services-core:uploadArchives ; \
-    export IS_LOCAL_DEVELOPMENT=true; ./gradlew :services-geojson:uploadArchives ; \
-    export IS_LOCAL_DEVELOPMENT=true; ./gradlew :services:uploadArchives ; \
-    export IS_LOCAL_DEVELOPMENT=true; ./gradlew :services-turf:uploadArchives ; \
+publish-to-bintray:
+	export IS_LOCAL_DEVELOPMENT=false; ./gradlew :services-core:bintrayUpload ; \
```
Comment (LukasPaczos) [inline Makefile]: `IS_LOCAL_DEVELOPMENT` is not used.

File: build.gradle (hunk 7 above)
Comment (langsmith) [reply build.gradle]: Removed

File: Makefile (hunk 9 above)
Comment (langsmith) [reply Makefile]: Removed

File: build.gradle (hunk 8 above)
Comment (langsmith) [reply build.gradle]: Removed

File: build.gradle (hunk 6 above)
Comment (langsmith) [reply build.gradle]: Removed

File: circle.yml (hunk 5 above)
Comment (langsmith) [reply circle.yml]: Fixed

**Comment by tobrun** [comment]
@langsmith tried creating a snapshot but hitting:

Task :artifactoryDeploy FAILED
Deploying artifact: http://oss.jfrog.org/oss-snapshot-local/com/mapbox/mapboxsdk/mapbox-sdk-core/4.9.2-SNAPSHOT/mapbox-sdk-core-4.9.2-SNAPSHOT.jar

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':artifactoryDeploy'.
> java.io.IOException: Failed to deploy file. Status code: 403 Response message: Artifactory returned the following errors:
 User mapbox-bot is not permitted to deploy 'com/mapbox/mapboxsdk/mapbox-sdk-core/4.9.2-SNAPSHOT/mapbox-sdk-core-4.9.2-SNAPSHOT.jar' into 'oss-snapshot-local:com/mapbox/mapboxsdk/mapbox-sdk-core/4.9.2-SNAPSHOT/mapbox-sdk-core-4.9.2-SNAPSHOT.jar'. Status code: 403

**Comment by langsmith** [comment]
Java SDK SNAPSHOTs should be working and available now @tobrun https://oss.jfrog.org/artifactory/oss-snapshot-local/com/mapbox/mapboxsdk/

Screen Shot 2019-07-09 at 5 55 58 PM


Screen Shot 2019-07-09 at 5 57 41 PM

**Pull request merged**